lib/internal/modules/esm/loader.js

```diff
@@ -237,13 +251,60 @@ class DefaultModuleLoader {
    * @param {string} parentURL Path of the parent importing the module.
    * @param {Record<string, string>} importAssertions Validations for the
    *                                                  module import.
-   * @returns {Promise<ExportedHooks | KeyedExports[]>}
+   * @returns {Promise<ExportedHooks>}
    *  A collection of module export(s) or a list of collections of module
    *  export(s).
    */
   async import(specifier, parentURL, importAssertions) {
+    const { specifierCache, serializedAttributes } =
+      this.#resolveCache.getSerialized(specifier, parentURL, importAssertions);
+    const removeCache = () => {
+      // Remove the cache entry if the import fails.
+      delete specifierCache[serializedAttributes];
+    };
+
+    // If there are no cache entry, create one:
+    if (specifierCache[serializedAttributes] == null) {
+      const result = this.#import(specifier, parentURL, importAssertions);
+      // Cache the Promise for now:
+      specifierCache[serializedAttributes] = result;
+      PromisePrototypeThen(result, (result) => {
+        // Once the promise has resolved, we can cache the ModuleJob itself.
+        specifierCache[serializedAttributes] = result;
+      }, removeCache);
+      return result;
+    }
+
+    // If the cache entry is a function, it's a static import that has already been successfully loaded:
+    if (typeof specifierCache[serializedAttributes] === 'function') {
+      return specifierCache[serializedAttributes] = specifierCache[serializedAttributes]();
+    }
+
+    // If the cached value is not a promise, it's already been successfully loaded:
+    if (!isPromise(specifierCache[serializedAttributes])) {
+      return specifierCache[serializedAttributes];
+    }
+
+    // If it's still a promise, we must have a fallback in case it fails:
+    const fallback = () => {
+      // If another fallback has already cached a promise, use this one:
+      if (specifierCache[serializedAttributes] != null) {
+        return PromisePrototypeThen(specifierCache[serializedAttributes], undefined, fallback);
+      }
+      // Otherwise create a new cache entry:
+      const result = this.#import(specifier, parentURL, importAssertions);
+      specifierCache[serializedAttributes] = result;
+      PromisePrototypeThen(result, undefined, removeCache);
+      return result;
+    };
+    return PromisePrototypeThen(specifierCache[serializedAttributes], undefined, fallback);
```

Comment:
Do we have test coverage for all of this logic?

Reply:
The test cases I added.